### DISCUSSION TITLE
format parameter for date fields

### DISCUSSION THREAD

**Discussion by ghost** [pydantic/pydantic#4295] [Feature Request]
It looks like the regex for parsing dates is hard-coded to the ISO standard: ` %Y-%m-%d`.  If you want to use USA human date format, `%d/%m/%Y`, you have to write a custom validator and use a clunky mechanism to reuse it for all your date fields.

Can the Field class for type, date, be enhanced to specify it like:

`my_date: date = Field(..., date_format=[r'(?P<year>\d{4})-(?P<month>\d{1,2})-(?P<day>\d{1,2})', r'(?P<day>\d{1,2})/(?P<month>\d{1,2})/(?P<year>\d{4})'])`

or, even better, using date parsing format strings:

`my_date: date = Field(..., date_format=['%d/%m/%Y', '%d/%m/%y'])`

**Comment by samuelcolvin** [comment]
Date parsing is changing in V2, but it still won't support this, in fact it would make supporting this much harder.

We might make validators easier to reuse, which should accomplish the same thing.

I don't think even now that it's particularly "clunky" to reuse logic, you can either use validator functions or a custom data type.

**Reply by ghost** [reply]
Hmm, write a custom date class so your date can be written as 01/03/2022.  Human language is, of course, second class to machine language.

I didn't mean that _reusing_ a validator is clunky.  I meant that the documented mechanism for doing so is clunky/verbose:
```
   my_date: date = Field(...)
    _never_referenced = validator('my_date', allow_reuse=True, pre=True)(validate_date)
```
With respect to making validators easier to reuse, why not,
```
   my_date: date = Field(..., pre_validator=validate_date)
```
What is the purpose of the, `allow_reuse` parameter?  What happens when the value is False?  Presumably, an error is generated because you accidentally reused a validator.  Why would you want that error?

**Reply by samuelcolvin** [reply]
`01/03/2022` is a terrible date format, I assume it's 1st March, but then you could be American in which case :exploding_head:, we have ISO8601 for a reason.

---

Yes, I think we'll support 

```py
my_date: date = Field(..., pre_validator=validate_date)
```
In future

---

`allow_reuse` exists because it's extremely easy to duplicate a validator name on a class, so we give an error on

```py
class Foobar(BaseModel):
    x: int
    y: int

    @validator('x')
    def check_int(cls, v):
        assert v != 42
        return v

    @validator('y')
    def check_int(cls, v):
        assert v != 24
        return v
```

At least that was the rationale originally I think, I'd be happy to consider removing it.

`allow_reuse` just removes this check.
